pkg/download/protocol.go

```diff
@@ -100,8 +102,30 @@ func (p *protocol) List(ctx context.Context, mod string) ([]string, error) {
 	if isRepoNotFoundErr && storageEmpty {
 		return nil, errors.E(op, errors.M(mod), errors.KindNotFound, goErr)
 	}
+	// we have to filter out pseudo versions to prevent following scenario:
+	// user does go get github.com/my/mod
+	// there is no sem-ver and the /list endpoint returns nothing - /latests endpoint gets hit
+	// Athens saves the pseudo version x1
+	// from now on every time user runs go get github.com/my/mod she/he will get pseudo version x1 even if a newer version x2 exists
+	// this is because /list returns non-empty list of versions (x1) and so /latest wont get hit
+	return union(goList, removePseudoVersions(strList)), nil
+}
+
+var pseudoVersionRE = regexp.MustCompile(`^v[0-9]+\.(0\.0-|\d+\.\d+-([^+]*\.)?0\.)\d{14}-[A-Za-z0-9]+(\+incompatible)?$`)
+
+func removePseudoVersions(allVersions []string) []string {
+	// copied from go cmd https://github.com/golang/go/blob/master/src/cmd/go/internal/modfetch/pseudo.go#L93
+	isPseudoVersion := func(v string) bool {
```

Comment:
nit: it might be a tiny memory saver to define it once in this package?

Reply:
Wouldn't the compiler optimize that? Anyway, I've removed the func entirely.